### PR TITLE
Core class cleanup

### LINKED ISSUE #356: Clean Up Core Class
Due to more growth in some core models than others (i.e. Out of Order) the abstract core class needs tidying up to ensure all functions and member variables are in the appropriate place

### LINKED ISSUE #386: `SimEng/src/lib/Elf.cc` missing cstdio
At least on Fedora 39 with GCC 13, `<cstdio>` is no longer brought in by other system headers.
A explicit `#include <cstdio>` is required for compilation, otherwise we get:

```
In file included from /home/tom/SimEng/src/lib/Elf.cc:1:
/home/tom/SimEng/src/include/simeng/Elf.hh:30:3: error: ‘uint32_t’ does not name a type
   30 |   uint32_t p_type;
      |   ^~~~~~~~
/home/tom/SimEng/src/include/simeng/Elf.hh:7:1: note: ‘uint32_t’ is defined in header ‘<cstdint>’; did you forget to ‘#include <cstdint>’?
    6 | #include "simeng/span.hh"
  +++ |+#include <cstdint>
    7 | 
```

### PR DESCRIPTION
**Pull request by FinnWilkinson**
Cleans up the Core classes by moving common member variables into the base class and ensuring the code aligns with the rest of the project.

Closes #356 
Fixes #386

### CODE REVIEW

File: src/include/simeng/arch/ProcessStateChange.hh [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#pragma once
```
Comment (jj16791) [inline src/include/simeng/arch/ProcessStateChange.hh:1]: How come this file needs no includes?

Comment (FinnWilkinson) [reply src/include/simeng/arch/ProcessStateChange.hh:1]: That is a great question.... perhaps when it is linked with the files that do use it, it has the necessary headers available?

It also seems to be ok in the Jenkin's pipeline, but I can add the headers it should need if preferable?

Comment (FinnWilkinson) [reply src/include/simeng/arch/ProcessStateChange.hh:1]: Headers added along with some general code "de-tangling"

**Review by jj16791 — DISMISSED** [review]

**Review by ABenC377 — CHANGES_REQUESTED** [review]
Only comment relates to the units used for clock frequency.  So far as I can tell, there isn't a good reason for using GHz in some places and Hz in others.  In which case, it would seem better to stick to one unit throughout the code base.  However, @FinnWilkinson @jj16791 @dANW34V3R @JosephMoore25, please let me know if I'm missing something

**Comment by jj16791** [comment]
> Only comment relates to the units used for clock frequency. So far as I can tell, there isn't a good reason for using GHz in some places and Hz in others. In which case, it would seem better to stick to one unit throughout the code base. However, @FinnWilkinson @jj16791 @dANW34V3R @JosephMoore25, please let me know if I'm missing something

You'd be correct in the use cases primarily converting the input GHz value into Hz. A significant reason behind the input being in GHz is that that's the natural unit of measure for modern HPC CPU cores. It's very unnatural to think about the frequency in terms of Hz and we'd need to add comments and notes to ensure that the user diverts from the natural assumptions. We kept it as GHz and got the code to convert it to Hz to make it "easier"/more convenient for the user.

**Comment by FinnWilkinson** [comment]
> > Only comment relates to the units used for clock frequency. So far as I can tell, there isn't a good reason for using GHz in some places and Hz in others. In which case, it would seem better to stick to one unit throughout the code base. However, @FinnWilkinson @jj16791 @dANW34V3R @JosephMoore25, please let me know if I'm missing something
> 
> You'd be correct in the use cases primarily converting the input GHz value into Hz. A significant reason behind the input being in GHz is that that's the natural unit of measure for modern HPC CPU cores. It's very unnatural to think about the frequency in terms of Hz and we'd need to add comments and notes to ensure that the user diverts from the natural assumptions. We kept it as GHz and got the code to convert it to Hz to make it "easier"/more convenient for the user.

Is there a reason for using Hz in the code at all though? We could store `clockFrequency_` as a float and in terms of GHz given the usage inside each core is to just divide it by 1e9 again (i.e. put it back from Hz to GHz)

File: src/include/simeng/Core.hh [hunk 2]
```diff
@@ -33,6 +50,69 @@ class Core {
...
+  /** The active exception handler. */
+  std::shared_ptr<arch::ExceptionHandler> exceptionHandler_;
+
+  /** The number of times this core has been ticked. */
+  uint64_t ticks_ = 0;
+
+  /** Whether or not the core has halted. */
+  bool hasHalted_ = false;
+
+  /** Clock frequency of core in Hz */
```
[the remark this answers is not in the document]
Comment (FinnWilkinson) [reply src/include/simeng/Core.hh:114]: Good point - I've updated this so that the clock frequency is stored as GHz inside the core

**Review by jj16791 — DISMISSED** [review]

**Review by ABenC377 — DISMISSED** [review]

**Review by JosephMoore25 — DISMISSED** [review]

**Review by dANW34V3R — DISMISSED** [review]

**Review by JosephMoore25 — APPROVED** [review]

**Review by jj16791 — APPROVED** [review]

**Review by dANW34V3R — DISMISSED** [review]

**Review by ABenC377 — APPROVED** [review]

**Review by jj16791 — APPROVED** [review]

**Review by JosephMoore25 — APPROVED** [review]

**Review by ABenC377 — APPROVED** [review]

**Pull request merged**